### ISSUE TITLE
[sig-release] Umbrella issue for a job that signs artifacts and uploads them to a GCS bucket

### ISSUE THREAD

**Issue by dims** [kubernetes/release#913]
Currently we need googlers to build, sign and upload deb/rpm artifacts. We need a prow job(s) that can do this. Ones that can be triggered by the sig-release team when they cut the release. 

- [ ] Find OWNERS for the deb and rpm definitions and associated scripts.
- [ ] Define policy for dependencies are named inside the deb/rpm definition metadata 
- [ ] We need a CNCF owned signing key. (We need to build a web of trust that signs this key, until then we can just use a temporary key, see kubernetes/kubernetes#70132)
- [ ] We need one or more GCS buckets for storing the artifacts (talk to wg-k8s-infra)
- [ ] Another GCS bucket for staging build artifacts pending approval
- [ ] We need a directory structure of how we would store the artifacts (to accomodate daily/nightly in addition to what we do today)
- [ ] Ability to Inspect staged artifacts (manually? automatically?) to ensure compliance with community-approved release process
- [ ] We need a trusted cluster for the job(s) (talk to wg-k8s-infra. The signing key will need to be loaded onto the cluster so the jobs can access it)
- [ ] We need a way to trigger the jobs (git-ops style, need a design for the yaml files, guessing we will need the SHA's for the k/k, k/release repos and version numbers)
- [ ] Ability for these jobs to Migrate approved release artifacts to GCS bucket for approved builds
- [ ] Add job definitions in test-infra to run the jobs in the trusted cluster

**Comment by justaugustus** [comment]
(Migrated to k/release)

/sig release
/area release-eng
/milestone v1.17
/priority important-soon
/kind feature cleanup

**Comment by nfisher** [comment]
If the packages are being PGP signed is it possible to place them on a HTTP host in addition to the HTTPS so that caching proxies can be used?

**Comment by justaugustus** [comment]
Will start working together w/ Arnaud on this in the coming months...
/assign @ameukam 
/milestone v1.24

**Comment by BenTheElder** [comment]
Note: Beyond getting builds run by community controlled infrastructure (instead of e.g. my workstation!) and published to GCS, we should also publish them to a proper package host with community controlled publishing AKA https://github.com/kubernetes/release/issues/913, I think that's the intent with https://github.com/kubernetes/release/issues/913#issuecomment-1015593512 but I don't see this called out in the issue body 😅

**Comment by ameukam** [comment]
Notes:
- `apt-key` is deprecated in favor of `signed-by` : https://wiki.debian.org/DebianRepository/UseThirdParty

**Comment by BenTheElder** [comment]
I want to highlight that today **there are just three active Googlers running the package build/publish on a volunteer basis, only two of which have experience doing so** and the process to do so has only gotten more onerous.

**Currently to this day even the package _builds_ happen on my workstation (!) instead of being controlled by the community**, which seems pretty unacceptable compared to the rest of our release process (which is largely in secure community owned cloud-builds without direct human access).

These packages are giving me hyperkube vibes ...

I would ask that we focus on getting these things under community management before we continue rethinking / rewriting the build tools etc.

We should be able to lift-and-shift builds to community build hosts and iterate from there.

**Comment by sftim** [comment]
If folks want to help move this work forward (and they aren't necessarily Googlers), how can they start to do so?

**Comment by BenTheElder** [comment]
Great question @sftim ❤️ 

Getting packages built and uploaded should be a conversation with SIG release, about how they'd like to move forward on this. It should not be blocked on Googlers at all.

Building packages artificially involves Googlers today, but the scripts to do so are in this repo and only use open source tools.

Additional scripts which are also open source in this repo then wrap those and continue on to add signatures to the packages and then publish them to the google package host, involving our internal signing and repo management tools.

I would say the approach that personally makes the most sense is:
- get familiar with the package build scripts
- `krel`, the kubernetes release tool, should start including a package build step running the build scripts
- once we know this works, the google build+publish+release script should be updated to download packages uploaded by the upstream release process (today it instead consumes k8s binaries and then packages them), it can still sign and publish them

In parallel or next:
- the project needs to setup a package host of some sort
- we will need to get users to switch. it will be at a new location and will have a new community managed signing key users will need to add, as we cannot export signing keys from the google infrastructure (literally, the key is not accessed, automation performs the signing by request, at no point do any of us have those keys)

The latter portion will also involve SIG K8s Infra, purely with respect to providing the necessary resources.

Artifact registry might be an option we could leverage: https://cloud.google.com/artifact-registry/docs/os-packages
... but at the moment SIG k8s Infra is over budget, which is what I'm working on currently (https://github.com/kubernetes/k8s.io/wiki/New-Registry-url-for-Kubernetes-(registry.k8s.io))


The top level script which does signing and publishing after invoking the builds: https://github.com/kubernetes/release/blob/51a27193f2adde4cbee55956c124d3eae1177fa0/hack/rapture/k8s-rapture.sh
Docs: https://github.com/kubernetes/release/blob/51a27193f2adde4cbee55956c124d3eae1177fa0/hack/rapture/README.md

The build scripts:

https://github.com/kubernetes/release/blob/51a27193f2adde4cbee55956c124d3eae1177fa0/packages/deb/jenkins.sh
https://github.com/kubernetes/release/blob/51a27193f2adde4cbee55956c124d3eae1177fa0/packages/rpm/docker-build.sh

(which you can see invoked in k8s-rapture.sh)

**Comment by ameukam** [comment]
> the project needs to setup a package host of some sort

I'll work on setup some sandbox infrastructure for this next week.

**Comment by LappleApple** [comment]
SIG Release has been working on building a plan to get this work over the finish line. Please review [this breakdown of related milestones and tasks](https://docs.google.com/spreadsheets/d/1Vgpb7HPg8WgoFtDuddaRD6j7jQLoqNK6vUu6Lb0ea28/edit#gid=885767119), which needs a bit more input and refinement. Then we should copy the items into a checklist posted directly in here, for visibility.

**Comment by sftim** [comment]
(from that spreadsheet)
> Risk: A new GPG key will be issued incompatible with the existing GPG key managed by Google. This is a breaking change for the existing infrastructures using DEB/RPM packages as installation artifacts.

I don't think the new key is a breaking change for people who have taken new action, or at least it doesn't need to be. If we use a new PGP key and _also_ change the repo URL from 
`https://packages.cloud.google.com/yum/repos/kubernetes-*` then folks with existing working setups can migrate. The breaking change to drop the old Google-linked repo can happen later.

I was assuming that the new repo URL would use a kubernetes.io domain (which might turn out to be a redirect).

**Comment by sftim** [comment]
If folks want to contribute to refining that plan, how can they do so?

**Comment by sftim** [comment]
(from that spreadsheet)
> Risk: A new GPG key will be issued incompatible with the existing GPG key managed by Google. This is a breaking change for the existing infrastructures using DEB/RPM packages as installation artifacts.

I don't think the new key is a breaking change for people who have taken new action, or at least it doesn't need to be. If we use a new PGP key and _also_ change the repo URL from 
`https://packages.cloud.google.com/yum/repos/kubernetes-*` then folks with existing working setups can migrate. The breaking change to drop the old Google-linked repo can happen later.

I was assuming that the new repo URL would use a kubernetes.io domain (which might turn out to be a redirect).

**Comment by saschagrunert** [comment]
@sftim feel free to add your suggestions directly to the spreadsheet since it is our source of truth for now. We plan to outline the whole topic in a KEP (as usual) and request input from others via the usual workflow. You can also add your name to one of the columns if you like to get directly involved in one of the milestone topics. :pray:

**Comment by BenTheElder** [comment]
ref: https://github.com/kubernetes/release/issues/913#issuecomment-1081121833

I'm going through off-boarding from the Kubernetes EngProd team @ Google and found that while I've been on vacation the past couple weeks **we dropped down to only two people in this rotation, and it will at least briefly be down to one** (⚠️) as I exit.

@MushuEE will be onboarding 1-2 junior team members O(soon) but that's where this is at on the Google end right now ...

I think the team is hiring, but it's relatively small right now, anything involving these packages just isn't prioritized with other more pressing work – they're more of a historical tech-debt handed off from the previous release-engineering team.

I'll do what I can to help, but my own priorities will be shifting / expanding with my [new role](https://twitter.com/BenTheElder/status/1551670320423059456), and I will be losing access to sign / publish to the Google infra.

**Comment by dims** [comment]
@BenTheElder thanks for the update. This is just an indication that we should increase the priority of this work and get it done quickly we've known this for a while now given that this issue has been open from 2019 ( cc @kubernetes/sig-release )

**Comment by BenTheElder** [comment]
I think folks in SIG release have started on this and I left a note in slack as well, there’s been recent activity like https://github.com/kubernetes/enhancements/pull/3434 🙏 that I haven’t caught up on yet, but I’m sure others are also watching this issue as well since we’ve linked to it from requests to change how signing is done etc 😅

**Comment by LappleApple** [comment]
+1 @dims. Here is the current work plan, with milestones broken down to highlight various needs and current progress noted. https://github.com/kubernetes/enhancements/issues/1731#issue-609450467 cc @detiber @ameukam @RobertKielty @Verolop (and @castrojo, who's offered to help do some outreach on behalf of the third-party hosting services-related tasks).

**Comment by sftim** [comment]
It sounds like the bit where folks _have_ to be a Googler, right now, is the package signing. Everything else can happen outside Google - this is an open source project after all. @dims has it right to focus on the signing aspect aside from the package build.

I wonder if a way to tackle this is to start with getting a Kubernetes-controlled private key, even a disposible one, to sign some trivial package. Maybe fetch an existing package using `curl`; maybe make a package that owns deploys `/hello-world.txt`

Once we can do that part, we can look at signing something more complex, and also how to better protect the private key.

**Comment by BenTheElder** [comment]
I believe the current google-internal hosting infrastructure is also coupled to the signing system, and because we're not even using the product offering there one way or another we're going to have to migrate both package signing and publishing/hosting in order to bring this to community control. 


Even the Google employees with permission to sign and publish packages cannot access, add, or modify the signing keys on the existing infrastructure.

In the short term, it's also time consuming that we conflate _building_ packages with access to sign and publish, since those are actually distinct steps, an intermediate step that could be worked on in parallel is moving the actual package build onto the community release process. Sometimes the package building is currently broken because it is in no way run in automation / CI, consuming even more time. The package sign and upload process is much quicker and less error-prone than the builds.

If we moved that to the upstream release build process, then the Googlers could download, then sign and publish those packages until the community signing and publishing is available. That would reduce the time required from that small rotation, and the community signing and packaging would also then have package builds available to work with. It also fixes the "these are built on someone's workstation and nobody can really vet them properly". After some transition period we could phase out publishing on the existing infra.

Previous discussion has been hampered by efforts to build an entirely new package build system, which we do not currently use at all. I would strongly urge that we lift-and-shift the existing build we ship releases with today and iteratively improve it later.

**Comment by saschagrunert** [comment]
@BenTheElder this is what we propose in https://github.com/kubernetes/enhancements/pull/3434, I’ll update the KEP today so that it’s ready for review.

**Comment by BenTheElder** [comment]
Thanks @saschagrunert, FWIW I did not read it that way: I read the KEP as blocked on replacing the package build tools as well. (Really glad to see that it's active regardless! 🙏)

**Comment by BenTheElder** [comment]
Actually, to immediately de-risk things somewhat at even lower cost we could simply do:
- split the rapture script into a build script and a sign + publish script
- enhance the build script to upload the outputs to a storage bucket
- enhance the sign + publish script to download from a storage bucket (we already input a k8s version)
- release team / release managers can take over running the build

The release team / managers the project at large can staff up, this two-person rotation we cannot.

I think the internal team may also have some automation options that are more viable to consider investing in as a stopgap to reduce some of the pressure on them and the other internal pushback if they don't have to get a functioning build environment with docker etc. ...

**Comment by saschagrunert** [comment]
@BenTheElder generelly yes, why not. :+1: 

> * release team / release managers can take over running the build

I see a bunch of pros and cons related to this and I get the overall intention of that take. Still I'd like to not increase the complexity on the release managers side by additionally running a _manual_ step during the release. Can we ensure that the package build script is that battle proof that it works on every release managers machine and produces the same outputs?

Alternatively, automating the rapture build script into `krel stage` seems to be a hack, because we explicitly decided to go away from running bash scripts in the release pipeline during the whole development of `krel` and deprecation of `anago`.

I think we should keep up the discussion and find a good way to integrate the changes faster without introducing unnecessary tech debt.

**Comment by saschagrunert** [comment]
@dims do we still need that umbrella issue with a possible merged of https://github.com/kubernetes/enhancements/pull/3434 and therefore an up to date KEP in https://github.com/kubernetes/enhancements/issues/1731?

**Comment by dims** [comment]
@saschagrunert nope. we can close it!

**Comment by saschagrunert** [comment]
Closing in favor of https://github.com/kubernetes/enhancements/issues/1731

**Comment by BenTheElder** [comment]
> I see a bunch of pros and cons related to this and I get the overall intention of that take. Still I'd like to not increase the complexity on the release managers side by additionally running a manual step during the release. Can we ensure that the package build script is that battle proof that it works on every release managers machine and produces the same outputs

I mean no more or less than it is for the _two_ part time volunteers that can only be googlers we block on today.

the requirements are well documented. Currently it requires docker and `rpm` and the script has been in use for ages with occasional minor updates.

As far as producing the same outputs ... again no more or less? Google workstations receive updates to docker and rpm too ... package building is not doing anything Google specific. Never has

**Comment by BenTheElder** [comment]
> Alternatively, automating the rapture build script into krel stage seems to be a hack, because we explicitly decided to go away from running bash scripts in the release pipeline during the whole development of krel and deprecation of anago.

This I don't quite understand either, `make release` in kubernetes is a metric ton of bash with the thinnest of makefile on top ... every krel pipeline is invoking a **lot** of bash 🙃

Invoking one more build script surely won't be a serious regression?

This package build code in the script is what we have shipped with since the first Kubernetes packages ever up through now, 8+ years into the project

**Comment by saschagrunert** [comment]
> > Alternatively, automating the rapture build script into krel stage seems to be a hack, because we explicitly decided to go away from running bash scripts in the release pipeline during the whole development of krel and deprecation of anago.
> 
> This I don't quite understand either, `make release` in kubernetes is a metric ton of bash with the thinnest of makefile on top ... every krel pipeline is invoking a **lot** of bash upside_down_face

I know, but we in SIG Release do not maintain that code directly on a daily basis.

> Invoking one more build script surely won't be a serious regression?

It will for us, because it moves the responsibility with the code. We decided against bash for the main reason that it's not being tested in the same way we can test golang applications. Means having the same logic in golang encapsulated in test-able units (a library) could be the right path forward for building the packages. We still evaluate OBS right now, so that discussion has to be finished before.

**Comment by BenTheElder** [comment]
> I know, but we in SIG Release do not maintain that code directly on a daily basis.

But, nobody has asked to maintain it. Invoking != maintaining, as evidenced by this conversation re: `make release`.

Further, using something for now does not mean it cannot be swapped out later or that there is a long term commitment to maintaining it.

Build, sign, publish are each a distinct step. Building is time heavy and should easily be executed in an automated build environment that has all the same requirements (`make release` has more environment dependencies actually).

---

As of this week, *anyone* can build the Debian and RPM packages Googlers have been building, signing and publishing since I've taken the straightforward step of factoring out the tiny build step into a distinct entrypoint: https://github.com/kubernetes/release/pull/2708 https://github.com/kubernetes/release/blob/master/hack/rapture/build-packages.sh

You need docker and bash installed. You clone the repo and run `hack/rapture/build-package.sh 1.25.2`, where `1.25.2` is the version of Kubernetes you wish to build packages for.

Any continued dependence on my former team having time to **build** packages is purely artificial, and does not need to be blocked on building a golang package build tool.


This script is not difficult to run, it is very little code to maintain, and I'm readily available to answer any questions on the subject, despite no special expertise being required.

I no longer have the access to publish these, but I'm doing my best to ensure that there is an opportunity to prevent these packages from going the way of hyperkube, if folks wish to take over. If not, there are many other options for installing Kubernetes and many other problems to work on. 

If I were invested in these distro packages I would reduce the dependency on the 1-2 people who have permission to publish these as quickly as possible.
